### ISSUE TITLE
CI needed, python 2 and python 3

### ISSUE THREAD

**Issue by balarsen** [spacepy/spacepy#7]
We should get CI setup, all the cool kids are doing it. 

What versions?
I use a bit of python2.7 and a lot of 3.6 (3.7 is also just out).

Are we still supporting 2.6 I forget?

**Comment by drsteve** [comment]
I've got Travis hooked in to a couple of my other repos - shouldn't be too hard to get it going for SpacePy. I'd say we should do 2.7, 3.5 and 3.6 at a minimum. I also don't think we need to keep supporting 2.6 since end-of-life for Python 2 is planned. Actually, I'm not sure that we still really do support 2.6, but if we do I'm happy to drop it. I guess we should document that.

**Comment by hugovk** [comment]
No need for Python 2.6, and some projects are planning or already dropping Python 2.7.

Here's the pip installs for SpacePy from PyPI for June 2018, showing no Python 2.6:

| python_version | percent | download_count |
| -------------- | ------: | -------------: |
| 3.6            |  44.44% |             24 |
| 2.7            |  40.74% |             22 |
| 3.5            |  12.96% |              7 |
| 3.7            |   1.85% |              1 |
| Total          |         |             54 |

Source: `pypinfo --start-date 2018-06-01 --end-date 2018-06-30 --percent --markdown SpacePy pyversion`

**Comment by balarsen** [comment]
For the CI we will want to install packages in a couple of ways to assess into the future the compatible versions. Will mean a kinda involved CI but its all in the background. Relates to #8.

For example:
numpy==1.7.0. #  make sure a version still works but also...
numpy>=1.7.0  #  to make sure it works on the most recent also

**Comment by drsteve** [comment]
Thanks. I know there's another ticket to cover documenting required versions, etc.

It's interesting to see the PyPI install stats. Certainly we still have some users who probably aren't installing from PyPI and @balarsen and I can probably check with known users to see whether 2.6 is still required by them.

I've assigned myself to this and I'll try to find time to set up Travis on my fork in the near future.

**Comment by drsteve** [comment]
Current pull request from my fork includes an initial Travis CI setup (3.5 and 3.6 only) and I've turned on the Travis GitHub app. Not closing this issue just yet as I was having some trouble with the Python 2.7 build... not test failures - it was freezing on trying to compile sources. The CI could/should also be expanded to test different versions of dependencies as @balarsen noted above.

It also looks like our consensus on version support is that 2.7 is still officially supported for the time being. As are 3.5, 3.6 (and we'll have to add 3.7). 2.6 is basically defunct at this point and doesn't need explicit support moving forward.

TL;DR:
- CI running, but leaving ticket open until 2.7 and specific dependency versions are tested.

**Comment by jtniehof** [comment]
What's the best entry point for learning the Travis integration, current status of the CI, etc? I'm seeing it in the PRs, which is nice, but are we running say nightly against master?

**Comment by drsteve** [comment]
Good questions. My entry point was not trivial... I had to take over managing the CI for another project that was a lot more complicated.

The Travis CI docs are not awful. [https://docs.travis-ci.com/user/getting-started/](https://docs.travis-ci.com/user/getting-started/)

Key things:
- Github has Travis CI integration (look under settings, integration and webhooks)
- Whenever there's a PR or a commit it triggers a run of the tests
- The test spec is in the .travis.yml file
- The usual way of including info on whether the project is passing is through a linked badge in the README.md (see, e.g, [PyForecastTools](https://github.com/drsteve/PyForecastTools) or [LANLGeoMag](https://github.com/drsteve/LANLGeoMag) )

There's almost certainly a way to do nightly builds, but I haven't looked into it. I think the more important thing (since the tests run every time there's a proposed change or an actual change) is testing against specific versions of dependent packages. We should probably close this and open new issues (maybe tagged as "enhancement" or something) asking for upgraded CI.

**Comment by jtniehof** [comment]
I can see fixing things by pinning the dependency version, but I also like the idea of having a failure giving us a heads-up that a new dependency version is giving us trouble. I will close this and open up some enhancements regarding 2.7, nighttly builds, pinned versions.

**Comment by hugovk** [comment]
For daily, weekly or monthly builds, go to the settings in the CI and look under cron jobs.

You can have it run always on schedule, or only if there's been no build on the last 24 hours. 

https://docs.travis-ci.com/user/cron-jobs/
